---
title: "RvE1 DO Shaikh Lab Study"
author: "Abrar Al-Shaer"
date: "March, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:/Users/Abrar/Dropbox/UNC_OneDrive/Shaikh Lab/DO Mice Project/UNC_Al-Shaer_MURGIGV01_20200710/DO Rqtl2 Analysis/")
```

Read in Data
```{r cars}
library(qtl2)
library(dlookr)
rm(list=ls())
contrl_file <- read_cross2("DO_study_SSH_qtl2_control_file.json")
summary(contrl_file)
names(contrl_file)
```

#Transformations

The QTL mapping model in qtl2 expects that the incoming data will be normally distributed. You may use transformations such as log or square root to make your data more normally distributed.

I already transformed my data beforehand based on the results below then recreated the control file with the newly transformed pheno data.
```{r}
#assessing distribution of weights in pheno data
hist(contrl_file$pheno[,2], main = "Weights")

#assessing distribution of glucose in pheno data
hist(contrl_file$pheno[,6], main = "Pre Glucose Values")
hist(log(contrl_file$pheno[,6]), main = "Log10 of Pre Glucose Values")
hist(contrl_file$pheno[,7], main = "Post Glucose Values")
hist(log(contrl_file$pheno[,7]), main = "Log10 of Post Glucose Values")

#assessing distribution of insulin in pheno data
hist(contrl_file$pheno[,8], main = "Pre Insulin Values")
hist(log(contrl_file$pheno[,8]), main = "Log10 of Pre Insulin Values")
hist(contrl_file$pheno[,13], main = "Post Insulin Values")
hist(log(contrl_file$pheno[,13]), main = "Log10 of Post Insulin Values")

#assessing distribution of GIP in pheno data
hist(contrl_file$pheno[,9], main = "Pre GIP Values")
hist(log(contrl_file$pheno[,9]), main = "Log10 of Pre GIP Values")
hist(contrl_file$pheno[,14], main = "Post GIP Values")
hist(log(contrl_file$pheno[,14]), main = "Log10 of Post GIP Values")

#assessing distribution of glucagon in pheno data
hist(contrl_file$pheno[,10], main = "Pre Glucagon Values")
hist(log(contrl_file$pheno[,10]), main = "Log10 of Pre Glucagon Values")
hist(contrl_file$pheno[,15], main = "Post Glucagon Values")
hist(log(contrl_file$pheno[,15]), main = "Log10 of Post Glucagon Values")

#assessing distribution of leptin in pheno data
hist(contrl_file$pheno[,11], main = "Pre Leptin Values")
hist(transform(contrl_file$pheno[,11], method = "sqrt"), main = "Sqrt of Pre Leptin Values")
hist(contrl_file$pheno[,16], main = "Post Leptin Values")
hist(transform(contrl_file$pheno[,16], method = "sqrt"), main = "Sqrt of Post Leptin Values")

#assessing distribution of resistin in pheno data
hist(contrl_file$pheno[,12], main = "Pre Resistin Values")
hist(transform(contrl_file$pheno[,12], method = "sqrt"), main = "Sqrt of Pre Resistin Values")
hist(contrl_file$pheno[,17], main = "Post Resistin Values")
hist(transform(contrl_file$pheno[,17], method = "sqrt"), main = "Sqrt of Post Resistin Values")

#same plots were made with the fold change variables (Glucose_FC, Insulin_FC, etc.). All fold changes were Log10 transformed.
```

#Creating a map of pseudomarkers
The first task in QTL analysis is to calculate conditional genotype probabilities, given the observed marker data, at each putative QTL position. The calc_genoprob() function calculates QTL genotype probabilities, conditional on the available marker data. These are needed for most of the QTL mapping functions. The result is returned as a list of three-dimensional arrays (one per chromosome). Each 3d array of probabilities is arranged as individuals x genotypes x positions. To find QTL at positions between markers (so called "pseudomarkers"), first insert pseudomarkers into the genetic map with the function insert_pseudomarkers().
```{r}
#step=1 means that pseudomarkers are spaced at 1 cM intervals from genotyped markers
map <- insert_pseudomarkers(map=contrl_file$gmap, step=1)

#compare structure of map vs gmap (to make sure they aren't the same, map should have more positions, # of positions is found in between the brackets after chr [])
str(map)
str(contrl_file$gmap)

length(map[["12"]]) #gives you the number of markers on chromosome 12
```

#Calculate QTL genotype probabilities
```{r}
pr <- calc_genoprob(contrl_file, map, error_prob=0.001, cores=4)
#names(pr)
#head(dimnames(pr$`19`))

#look at a marker location for first 3 samples
(pr$`19`)[1:3,,"UNCHS047438"]
```

Plot genotype probabilities
```{r}
#ind: the index of the individual to plot
plot_genoprob(pr, map, ind = 1, chr = 19)
```

#Kinship Matrix
Calculating Kinship Matrix
```{r}
kinship_loco <- calc_kinship(probs = pr, type = "loco", cores = 4)
```

Visualize kinship matrix
```{r}
n_samples <- 50 #250
heatmap(kinship_loco[[1]][1:n_samples, 1:n_samples]) #symm = TRUE
```

#Covariates
We will create additive covariates that will be used in the mapping model. We will use study cohort as a covariate in the mapping model. Our study contained only male mice, but in most cases, you would include sex as an additive covariate as well.
```{r}
#forms a matrix that is fit to go into a model
#it'll go into the covar dataframe and take the cohort column and create a matrix
#if you have multiple covariates you can write ~cohort+sex+diet...etc.

addcovar <- model.matrix(~cohort+Obesity, data = data.frame(contrl_file$covar))[,-1]
#[,-1] is there b/c we must delete the first intercept column because we don't need it we just want to get the encoded covariates in a matrix
```


#Genome Scan
Genome scans for glucose, insulin, glucagon, GIP, leptin, & resistin 
```{r}
#index <- which(colnames(contrl_file$pheno) == "Log10.Glucose.FC")
qtl_glucose <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Glucose.FC"], kinship = kinship_loco, addcovar = addcovar)
qtl_insulin <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Insulin_FC"], kinship = kinship_loco, addcovar = addcovar)
qtl_gip <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.GIP_FC"], kinship = kinship_loco, addcovar = addcovar)
qtl_glucagon <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Glucagon_FC"], kinship = kinship_loco, addcovar = addcovar)
qtl_leptin <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Leptin_FC"], kinship = kinship_loco, addcovar = addcovar)
qtl_resistin <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Resistin_FC"], kinship = kinship_loco, addcovar = addcovar)

qtl_weights <- scan1(genoprobs = pr, pheno = log10(contrl_file$pheno[,"Weight_g"]), kinship = kinship_loco, addcovar = addcovar)
qtl_weights_delta <- scan1(genoprobs = pr, pheno = log10(contrl_file$pheno[,"Delta.Weight.FC"]), kinship = kinship_loco, addcovar = addcovar)
qtl_fat <- scan1(genoprobs = pr, pheno = log10(contrl_file$pheno[,"Fat.mass_g"]), kinship = kinship_loco, addcovar = addcovar)
qtl_lean <- scan1(genoprobs = pr, pheno = log10(contrl_file$pheno[,"Lean.mass_g"]), kinship = kinship_loco, addcovar = addcovar)
qtl_fat_tertiles <- scan1(genoprobs = pr, pheno = contrl_file$pheno[,"Tertiles_Fat_Mass"], kinship = kinship_loco, addcovar = addcovar)
```

Plot the genome scan results
```{r}
plot_scan1(x = qtl_glucose, map = map, main = "Glucose Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_insulin, map = map, main = "Insulin Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_gip, map = map, main = "GIP Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_glucagon, map = map, main = "Glucagon Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_leptin, map = map, main = "Leptin Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_resistin, map = map, main = "Resistin Fold Changes, Pre/Post RvE1 Injection")
plot_scan1(x = qtl_weights, map = map, main = "Weights")
plot_scan1(x = qtl_weights_delta, map = map, main = "Delta Weights")
plot_scan1(x = qtl_fat, map = map, main = "Fat Mass")
plot_scan1(x = qtl_lean, map = map, main = "Lean Mass")
plot_scan1(x = qtl_fat_tertiles, map = map, main = "Tertiles Fat Mass")
```

#Performing a permutation test
The perms object contains the maximum LOD score from each genome scan of permuted data.
```{r}
#perms <- scan1perm(genoprobs = pr, pheno = contrl_file$pheno[,"Log10.Leptin_FC"], addcovar = addcovar, n_perm = 100, cores = 4)

#hist(perms)
#thr = summary(perms)
#summary(perms)

#plot_scan1(qtl_i,map)
#add_threshold(map, thr, col="red")

#peaks = find_peaks(qtl_i, map, threshold = thr, prob = 0.95)
#print("Significant LOD Score Peaks")
#peaks

cat("LOD QTLs Permutation Tests Results:\n

> summary(perms_insulin)
LOD thresholds (1000 permutations)
     pheno1
0.05     28
**NO significance**\n

> summary(perms_leptin, alpha = c(0.1, 0.05))
LOD thresholds (100 permutations)
     pheno1
0.1    35.6
0.05   36.2
**NO significance**\n

summary(perms_glucose, alpha = c(0.1, 0.05))
LOD thresholds (100 permutations)
     pheno1
0.1    20.6
0.05   21.1
**NO significance**\n

> summary(perms_gip, alpha = c(0.1, 0.05))
LOD thresholds (100 permutations)
     pheno1
0.1    26.1
0.05   26.6
**NO significance**\n

> summary(perms_glucagon, alpha = c(0.1, 0.05))
LOD thresholds (100 permutations)
     pheno1
0.1    21.3
0.05   22.0
**NO significance**\n

> summary(perms_resistin, alpha = c(0.1, 0.05))
LOD thresholds (100 permutations)
     pheno1
0.1    22.6
0.05   23.1
**NO significance**\n

*I tried adjusting by fat mass, and by 'obesity' (binary yes (14+ g fat mass) / no (<14g fat mass) - made no difference.")
```

Weights QTL Analysis
The support interval is determined using the Bayesian Credible Interval and represents the region most likely to contain the causative polymorphism(s). We can obtain this interval by adding a prob argument to find_peaks. We pass in a value of 0.95 to request a support interval that contains the causal SNP 95% of the time.
```{r}
#Permutation Analysis
perms100 <- scan1perm(genoprobs = pr, pheno = log10(contrl_file$pheno[,"Weight_g"]), addcovar = addcovar, n_perm = 100, cores = 4)
thr = summary(perms100)
summary(perms100, alpha = c(0.1, 0.05))
peaks = find_peaks(qtl_weights, map, threshold = thr, prob = 0.95)
peaks
print("From the peaks output above, the location of the maximum LOD score is at 33.128 Mb.")

#Plot QTLs
plot_scan1(x = qtl_weights, map = map, main = "Weights")
add_threshold(map, thr, col="red")

#png("LOD_scores_plot.png", units="in", width=6, height=5, res=600) 

#  lodindex lodcolumn chr    pos      lod  ci_lo  ci_hi
#1        1    pheno1   4 33.082 21.94957 32.901 33.128
#2        1    pheno1   7 36.359 20.02021 36.309 37.048
#3        1    pheno1  10 50.049 17.78711 49.758 50.157
#4        1    pheno1  13 53.109 18.45462 52.977 64.772
#5        1    pheno1  18 42.661 18.06568 42.593 42.913
```

#Estimated QTL effects
We will now zoom in on Chr 4 and look at the contribution of each of the eight founder alleles to the post-HF diet weight. Remember, the mapping model above estimates the effect of each of the eight DO founders. We can plot these effects (also called 'coefficients') across Chr 4 using scan1coef.
```{r}
chr = peaks$chr[1]
start = peaks$ci_lo
end = peaks$ci_hi
coef4 <- scan1blup(genoprobs = pr[,chr], pheno = log10(contrl_file$pheno[,"Weight_g"]), kinship <- kinship_loco[[chr]], addcovar = addcovar, cores = 4) #BLUP = best unbiased linear predictors
```

This produces an object containing estimates of each of the eight DO founder allele effect. These are the beta(j) values in the mapping equation.

The top panel in the plot below shows the eight founder allele effects (or model coefficients) along Chr 4. The founder allele effects are centerd at zero and the units are the same as the phenotype. You can see that DO mice containing the WSB/EiJ allele near 33 Mb have lower levels of weight gain. This means that the WSB allele is associated with less weight gain after high fat feeding, therefore may indicate a protective effect. The bottom panel shows the LOD score, with the support interval for the peak colored blue.
You can also see that the locus at ~55 Mb with the red PWK allele is trending towards significance.
```{r}
plot_coefCC(x = coef4, map = map, scan1_output = qtl_weights, main = "15-Week HF Diet Weight Gain Effects", legend = "bottomright") #colors by CC lines

#png("QTL_effects_plot.png", units="in", width=6, height=5, res=600) 
```

#SNP Association Mapping
We can call scan1snps to perform association mapping in the QTL interval on Chr 4. We first create variables for the chromosome and support interval where we are mapping. We then create a function to get the SNPs from the founder SNP database. The path to the SNP database (snpdb_file argument) points to the data directory on your computer. Note that it is important to use the keep_all_snps = TRUE in order to return all SNPs.
```{r}
query_func <- create_variant_query_func("C:/Users/Abrar/Dropbox/UNC_OneDrive/Shaikh Lab/DO Mice Project/UNC_Al-Shaer_MURGIGV01_20200710/DO Rqtl2 Analysis/cc_variants.sqlite") #directory to CC SNPs database

assoc <- scan1snps(genoprobs = pr[,chr], map = map, pheno = log10(contrl_file$pheno[,"Weight_g"]), kinship = kinship_loco, addcovar = addcovar, query_func = query_func, chr = chr, start = start, end = end, keep_all_snps = TRUE)
```

Plot SNP Association Map
This plot shows the LOD score for each SNP in the QTL interval. The SNPs occur in "shelves" because all of the SNPs in a haplotype block have the same founder strain pattern. The SNPs with the highest LOD scores are the ones for which WSB/EiJ contributes the alternate allele.
```{r}
plot_snpasso(scan1output = assoc$lod, snpinfo = assoc$snpinfo, main = "15-Week HF Diet Weight Gain Effects")
```

Add genes to association plot
```{r}
#mgi = mouse genome informatics
gene_func = create_gene_query_func("C:/Users/Abrar/Dropbox/UNC_OneDrive/Shaikh Lab/DO Mice Project/UNC_Al-Shaer_MURGIGV01_20200710/DO Rqtl2 Analysis/mouse_genes_mgi.sqlite")
genes = gene_func(chr, start, end)
plot_snpasso(assoc$lod, assoc$snpinfo, drop_hilit=1.5, genes = genes, top_panel_prop = 0.5, main = "15-Week HF Diet Weight Gain QTL LODs & Genes")

#top_panel_prop = We can adjust the proportion of the plot allocated for SNPs (on the top) and genes (on th bottom) using the top_panel_prop argument.
#drop_hilit = highlighting the top SNPs in the SNP association plot using the drop_hilit argument. SNPs with LOD score within a 1.5 LOD peak of the maximum SNPs (max - 1.5) are shown in pink

#png("QTL_effects_plot_w-Genes.png", units="in", width=7, height=5, res=600)
```

Get the top SNPs
In the table below, 1 = major allele, 2 = minor allele, 3 = rare 3 letter combination
```{r}
top = top_snps(assoc$lod, assoc$snpinfo) #the defaul is to get everything within a 1.5 LOD peak of the maximum SNPs (max - 1.5) max minus 1.5
head(top)
#write.csv(top, "top_snps_output_DO_Weights_SNPs_Analysis.csv")
```
The table results above show that AJ has 45 "2"s - i.e. minor allele SNPs, 129S1_SvImJ has 46 minor alleles, followed by NOD with 17, CAST with 14, and PWK with 15. 
